tests/enterprise-contract/contract.go

```diff
@@ -0,0 +1,146 @@
+
+	BeforeAll(func() {
+		fwk, err = framework.NewFramework(constants.TEKTON_CHAINS_E2E_USER)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(fwk.UserNamespace).NotTo(BeNil(), "failed to create sandbox user")
+		namespace = fwk.UserNamespace
+		kubeController = tekton.KubeController{
+			Commonctrl: *fwk.AsKubeAdmin.CommonController,
+			Tektonctrl: *fwk.AsKubeAdmin.TektonController,
+			Namespace:  namespace,
+		}
+		publicKey, err := kubeController.GetTektonChainsPublicKey()
+		Expect(err).ToNot(HaveOccurred())
+		GinkgoWriter.Printf("Copy public key from %s/signing-secrets to a new secret\n", constants.TEKTON_CHAINS_NS)
+		Expect(kubeController.CreateOrUpdateSigningSecret(
+			publicKey, publicSecretName, namespace)).To(Succeed())
+
+		defaultEcp, err := kubeController.GetEnterpriseContractPolicy("default", "enterprise-contract-service")
+		Expect(err).NotTo(HaveOccurred())
+		policySource = defaultEcp.Spec.Sources
+
+		cm, err := kubeController.Commonctrl.GetConfigMap("ec-defaults", "enterprise-contract-service")
+		Expect(err).ToNot(HaveOccurred())
+		verifyECTaskBundle = cm.Data["verify_ec_task_bundle"]
+		Expect(verifyECTaskBundle).ToNot(BeEmpty())
+		GinkgoWriter.Printf("Using verify EC task bundle: %s\n", verifyECTaskBundle)
+
+	})
+
+	Context("Release Policy", func() {
+		BeforeAll(func() {
+			generator = tekton.VerifyEnterpriseContract{
+				Bundle:              verifyECTaskBundle,
+				Image:               imageWithDigest,
+				Name:                "verify-enterprise-contract",
+				Namespace:           namespace,
+				PolicyConfiguration: "ec-policy",
+				PublicKey:           fmt.Sprintf("k8s://%s/%s", namespace, publicSecretName),
+				SSLCertDir:          "/var/run/secrets/kubernetes.io/serviceaccount",
+				Strict:              false,
+			}
+			secretName := fmt.Sprintf("golden-image-public-key%s", util.GenerateRandomString(10))
+			//The staging public key for verificaiton image
+			publicKey := []byte("-----BEGIN PUBLIC KEY-----\n" +
+				"MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEZP/0htjhVt2y0ohjgtIIgICOtQtA\n" +
+				"naYJRuLprwIv6FDhZ5yFjYUEtsmoNcW7rx2KM6FOXGsCX3BNc7qhHELT+g==\n" +
+				"-----END PUBLIC KEY-----")
+
+			GinkgoWriter.Println("Create golden image public signing key")
+			Expect(kubeController.CreateOrUpdateSigningSecret(publicKey, secretName, namespace)).To(Succeed())
+			generator.PublicKey = fmt.Sprintf("k8s://%s/%s", namespace, secretName)
+			pipelineRunTimeout = int(time.Duration(5) * time.Minute)
+
+		})
+
+		It("verifies the release policy: Task bundle is not acceptable", func() {
+			policy := ecp.EnterpriseContractPolicySpec{
+				Sources: policySource,
+				Configuration: &ecp.EnterpriseContractPolicyConfiguration{
+					Include: []string{"attestation_task_bundle.unacceptable_task_bundle"},
+				},
+			}
+			Expect(kubeController.CreateOrUpdatePolicyConfiguration(namespace, policy)).To(Succeed())
+
+			generator.Image = "quay.io/redhat-appstudio/ec-golden-image:08b938499ee47a25673f5e6a56f69ce57d87b45c"
```

Comment:
I wonder if we should tag this image with something more meaningful so it's not accidentally deleted? I think I have access to do so. Let me know what you think!

Reply:
> I wonder if we should tag this image with something more meaningful so it's not accidentally deleted? I think I have access to do so. Let me know what you think!

Do you mean generate new images with meaningful tag name or only update the existed ones' tag name?

Reply:
Only create a new tag for the same image, e.g. `quay.io/redhat-appstudio/ec-golden-image:e2e-test-unacceptable-task`.

At some point, someone may want to clean up old images in that repo. A tag like this might make it more obvious that a certain image should not be deleted.

Reply:
I did some try on this. 
1. built a image named quay.io/cuipinghuo/golden-container:f5e5f21faf47b199e3b596d67b0291214768af2c5af6322b5b34c53dd68ef0e1, then 
```
$ ec validate image --image quay.io/cuipinghuo/golden-container:f5e5f21faf47b199e3b596d67b0291214768af2c5af6322b5b34c53dd68ef0e1 --public-key k8s://tekton-chains/signing-secrets --policy tekton-chains/ec-policy | jq
{
  "success": false,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "quay.io/cuipinghuo/golden-container@sha256:f2fcad686dc32bb1cafa40026e23b3e8777ca05b62d825af0e57a6a89382f8e8",
      "violations": [
        {
          "msg": "Pipeline task 'build-container' uses an unacceptable task bundle 'quay.io/redhat-appstudio-tekton-catalog/task-buildah:0.1@sha256:c3712257615d206ef40013bf1c5c681670fc8f7fd6aac9fa4c86f7afeff627ef'",
          "metadata": {
            "code": "attestation_task_bundle.unacceptable_task_bundle",
            "effective_on": "2022-01-01T00:00:00Z"
          }
        },
        {
          "msg": "Pipeline task 'init' uses an unacceptable task bundle 'quay.io/redhat-appstudio-tekton-catalog/task-init:0.1@sha256:5ce77110e2a49407a69a7922042dc0859f7e8f5f75dc0cd0bcc2d17860469bdb'",
          "metadata": {
            "code": "attestation_task_bundle.unacceptable_task_bundle",
            "effective_on": "2022-01-01T00:00:00Z"
          }
        }
      ],
```
2. re-tag the image to quay.io/cuipinghuo/golden-container:e2e-test-unacceptable-task
```
$ sudo docker pull quay.io/cuipinghuo/golden-
$ sudo docker tag 8943e9265a8e quay.io/cuipinghuo/golden-container:e2e-test-unacceptable-task
$ sudo docker push quay.io/cuipinghuo/golden-container:e2e-test-unacceptable-task
```
3. validate the new tagged image
```
$ ec validate image --image quay.io/cuipinghuo/golden-container:e2e-test-unacceptable-task --public-key k8s://tekton-chains/signing-secrets --policy tekton-chains/ec-policy | jq
{
  "success": false,
  "components": [
    {
      "name": "Unnamed",
      "containerImage": "quay.io/cuipinghuo/golden-container@sha256:bccfca8867955aa274ff61bd3b0be6fc017febf04fe012784e95d29c50226180",
      "violations": [
        {
          "msg": "Image signature check failed: no signatures found for image"
        },
        {
          "msg": "No image attestations found matching the given public key. Verify the correct public key was provided, and one or more attestations were created."
        }
      ],
      "success": false
    }
  ],
```

Is this your way to create a new tag for the same image?

Reply:
Yes, that's one way to do it. Another is to simply use skopeo:
```
skopeo copy \
  docker://quay.io/cuipinghuo/golden-container:f5e5f21faf47b199e3b596d67b0291214768af2c5af6322b5b34c53dd68ef0e1 \
  docker://quay.io/cuipinghuo/golden-container:e2e-test-unacceptable-task
```

I can run a similar command on the `quay.io/redhat-appstudio/ec-golden-image`. Just let me know!

Reply:
Thanks @lcarva 
My point is new tagged quay.io/cuipinghuo/golden-container:e2e-test-unacceptable-task will fail ec validate with ```
"violations": [
        {
          "msg": "Image signature check failed: no signatures found for image"
        },
        {
          "msg": "No image attestations found matching the given public key. Verify the correct public key was provided, and one or more attestations were created."
        }
      ],
```

Reply:
Oh of course. You have to use `cosign copy`, so it copies the signatures and attestations.

Reply:
Wow, cosign is so cool.
Could you help me new tag `quay.io/redhat-appstudio/ec-golden-image:08b938499ee47a25673f5e6a56f69ce57d87b45c` and `quay.io/redhat-appstudio/ec-golden-image:c2f9e02e272ff8da32133ebcff454560c9dd0e38`? Thanks!

```
$ cosign copy quay.io/redhat-appstudio/ec-golden-image:08b938499ee47a25673f5e6a56f69ce57d87b45c quay.io/redhat-appstudio/ec-golden-image:e2e-test-unacceptable-task
$ cosign copy quay.io/redhat-appstudio/ec-golden-image:c2f9e02e272ff8da32133ebcff454560c9dd0e38
 quay.io/redhat-appstudio/ec-golden-image:e2e-test-out-of-date-task
```

Reply:
All set!

```
🐚 cosign tree quay.io/redhat-appstudio/ec-golden-image:e2e-test-out-of-date-task
📦 Supply Chain Security Related artifacts for an image: quay.io/redhat-appstudio/ec-golden-image:e2e-test-out-of-date-task
└── 💾 Attestations for an image tag: quay.io/redhat-appstudio/ec-golden-image:sha256-6858e6cb3ccf9caadf79c30f679a26f2daeb84b999b77561a3e9000d9b983441.att
   ├── 🍒 sha256:b0510cb023c91c90df16595915ec1180039c94880cdcc278b4cc9ff68af4d8c8
   └── 🍒 sha256:399147f4350cd7a6e1f0cd1589f7a48fd68ab494f041317c27389d2a5ecc2f28
└── 🔐 Signatures for an image tag: quay.io/redhat-appstudio/ec-golden-image:sha256-6858e6cb3ccf9caadf79c30f679a26f2daeb84b999b77561a3e9000d9b983441.sig
   └── 🍒 sha256:fa758aaf821d43ea323ed5d9bb937ef5bfce8e84f2101eb872b19627f90f6a24
└── 📦 SBOMs for an image tag: quay.io/redhat-appstudio/ec-golden-image:sha256-6858e6cb3ccf9caadf79c30f679a26f2daeb84b999b77561a3e9000d9b983441.sbom
   └── 🍒 sha256:963b025cf1988ec6945ec21e830bca7792aa20f34aa9f4287e40fafb404fdca7

🐚 cosign tree quay.io/redhat-appstudio/ec-golden-image:e2e-test-unacceptable-task
📦 Supply Chain Security Related artifacts for an image: quay.io/redhat-appstudio/ec-golden-image:e2e-test-unacceptable-task
└── 🔐 Signatures for an image tag: quay.io/redhat-appstudio/ec-golden-image:sha256-304040ca1911aa4d911bd7c6d6d07193c57dc49dbc43e63828b42ab204fb1b25.sig
   └── 🍒 sha256:62d4934673ad65910761251e0285dd9a96efecb55ea1fc484c53136f348f3ec1
└── 📦 SBOMs for an image tag: quay.io/redhat-appstudio/ec-golden-image:sha256-304040ca1911aa4d911bd7c6d6d07193c57dc49dbc43e63828b42ab204fb1b25.sbom
   └── 🍒 sha256:a20b9fe275f99a2b8d3402413f3d6af7cc6b1893d7ee706c42f14614e1420f31
└── 💾 Attestations for an image tag: quay.io/redhat-appstudio/ec-golden-image:sha256-304040ca1911aa4d911bd7c6d6d07193c57dc49dbc43e63828b42ab204fb1b25.att
   ├── 🍒 sha256:5213f952ae5e0ef0205c5d66ffebb43b395a1352fe65ae910c3adfda22966734
   └── 🍒 sha256:67041021df85bc1a4bd66ee80756043ef934fab58a62149d18be3d0cc07b2263
```